# Hotel Booking Data Analysis

Bu layihədə Python və Pandas vasitəsilə 119,390 otel rezervasiyasından ibarət dataset üzərində məlumatların təmizlənməsi, yeni göstəricilərin yaradılması və exploratory data analysis aparmışam.

Analizdə rezervasiyaların ləğv statusu, lead time, ADR, otel növü, müştəri tipi, gecələmə müddəti və hesablanmış rezervasiya dəyəri arasındakı əlaqələri araşdırmışam.


In [3]:
import pandas as pd

In [4]:
df = pd.read_csv(r"Data/Hotel_Bookings.csv")

In [5]:
# EDA - Exploratory Data Analysis
# Sətir və sütunların sayını göstərir.
df.shape

(119390, 32)

In [6]:
# Data Cleaning
# Bütün sütunları null (NaN) olan, yəni boş olan sətirləri silirik. 
df.dropna(how="all", inplace=True)

In [40]:
# Data Cleaning
# Datada olan sütunlardakı null sətirlərin faiz nisbəti
(df.isnull().sum()/len(df))*100

hotel                              0.000000
is_canceled                        0.000000
lead_time                          0.000000
arrival_date_year                  0.000000
arrival_date_month                 0.000000
arrival_date_week_number           0.000000
arrival_date_day_of_month          0.000000
stays_in_weekend_nights            0.000000
stays_in_week_nights               0.000000
adults                             0.000000
children                           0.003350
babies                             0.000000
meal                               0.000000
country                            0.408744
market_segment                     0.000000
distribution_channel               0.000000
is_repeated_guest                  0.000000
previous_cancellations             0.000000
previous_bookings_not_canceled     0.000000
reserved_room_type                 0.000000
assigned_room_type                 0.000000
booking_changes                    0.000000
deposit_type                    

In [8]:
# Data Cleaning
# Datada olan sütunlardakı null sətirlərin cəmi
df.isnull().sum()

hotel                                  0
is_canceled                            0
lead_time                              0
arrival_date_year                      0
arrival_date_month                     0
arrival_date_week_number               0
arrival_date_day_of_month              0
stays_in_weekend_nights                0
stays_in_week_nights                   0
adults                                 0
children                               4
babies                                 0
meal                                   0
country                              488
market_segment                         0
distribution_channel                   0
is_repeated_guest                      0
previous_cancellations                 0
previous_bookings_not_canceled         0
reserved_room_type                     0
assigned_room_type                     0
booking_changes                        0
deposit_type                           0
agent                              16340
company         

In [9]:
# EDA - Exploratory Data Analysis
# dtype-lafrını göstərir
df.dtypes

hotel                                 str
is_canceled                         int64
lead_time                           int64
arrival_date_year                   int64
arrival_date_month                    str
arrival_date_week_number            int64
arrival_date_day_of_month           int64
stays_in_weekend_nights             int64
stays_in_week_nights                int64
adults                              int64
children                          float64
babies                              int64
meal                                  str
country                               str
market_segment                        str
distribution_channel                  str
is_repeated_guest                   int64
previous_cancellations              int64
previous_bookings_not_canceled      int64
reserved_room_type                    str
assigned_room_type                    str
booking_changes                     int64
deposit_type                          str
agent                             

In [10]:
# EDA - Exploratory Data Analysis
# Datanın hansı sətirlərində nə qədər null olub olmadığını göstərir.
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 119390 entries, 0 to 119389
Data columns (total 32 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   hotel                           119390 non-null  str    
 1   is_canceled                     119390 non-null  int64  
 2   lead_time                       119390 non-null  int64  
 3   arrival_date_year               119390 non-null  int64  
 4   arrival_date_month              119390 non-null  str    
 5   arrival_date_week_number        119390 non-null  int64  
 6   arrival_date_day_of_month       119390 non-null  int64  
 7   stays_in_weekend_nights         119390 non-null  int64  
 8   stays_in_week_nights            119390 non-null  int64  
 9   adults                          119390 non-null  int64  
 10  children                        119386 non-null  float64
 11  babies                          119390 non-null  int64  
 12  meal                       

In [11]:
# EDA - Exploratory Data Analysis
# Sütunların adlarını göstərir.
df.columns

Index(['hotel', 'is_canceled', 'lead_time', 'arrival_date_year',
       'arrival_date_month', 'arrival_date_week_number',
       'arrival_date_day_of_month', 'stays_in_weekend_nights',
       'stays_in_week_nights', 'adults', 'children', 'babies', 'meal',
       'country', 'market_segment', 'distribution_channel',
       'is_repeated_guest', 'previous_cancellations',
       'previous_bookings_not_canceled', 'reserved_room_type',
       'assigned_room_type', 'booking_changes', 'deposit_type', 'agent',
       'company', 'days_in_waiting_list', 'customer_type', 'adr',
       'required_car_parking_spaces', 'total_of_special_requests',
       'reservation_status', 'reservation_status_date'],
      dtype='str')

In [12]:
# Data Cleaning
# "reservation_status_date" sütunu tarixdir,lakin object kimi qeyd olunub, onu tarixə çevirmək üçün bunu edirik:
df["reservation_status_date"] = pd.to_datetime(df["reservation_status_date"], errors="coerce")

In [13]:
# EDA - Exploratory Data Analysis
# Datanın hansı sətirlərində nə qədər null olub olmadığını təzədən yoxlayırıq
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 119390 entries, 0 to 119389
Data columns (total 32 columns):
 #   Column                          Non-Null Count   Dtype         
---  ------                          --------------   -----         
 0   hotel                           119390 non-null  str           
 1   is_canceled                     119390 non-null  int64         
 2   lead_time                       119390 non-null  int64         
 3   arrival_date_year               119390 non-null  int64         
 4   arrival_date_month              119390 non-null  str           
 5   arrival_date_week_number        119390 non-null  int64         
 6   arrival_date_day_of_month       119390 non-null  int64         
 7   stays_in_weekend_nights         119390 non-null  int64         
 8   stays_in_week_nights            119390 non-null  int64         
 9   adults                          119390 non-null  int64         
 10  children                        119386 non-null  float64       
 11

In [14]:
# EDA - Exploratory Data Analysis
# DataFrame-dəki ədədi (numeric) sütunlar haqqında ümumi statistik məlumat.
df.describe()

,is_canceled,lead_time,arrival_date_year,arrival_date_week_number,arrival_date_day_of_month,stays_in_weekend_nights,stays_in_week_nights,adults,children,babies,...,previous_cancellations,previous_bookings_not_canceled,booking_changes,agent,company,days_in_waiting_list,adr,required_car_parking_spaces,total_of_special_requests,reservation_status_date
count,119390.000000,119390.000000,119390.000000,119390.000000,119390.000000,119390.000000,119390.000000,119390.000000,119386.000000,119390.000000,...,119390.000000,119390.000000,119390.000000,103050.000000,6797.000000,119390.000000,119390.000000,119390.000000,119390.000000,119390
mean,0.370416,104.011416,2016.156554,27.165173,15.798241,0.927599,2.500302,1.856403,0.103890,0.007949,...,0.087118,0.137097,0.221124,86.693382,189.266735,2.321149,101.831122,0.062518,0.571363,2016-07-30 00:24:47.883407
min,0.000000,0.000000,2015.000000,1.000000,1.000000,0.000000,0.000000,0.000000,0.000000,0.000000,...,0.000000,0.000000,0.000000,1.000000,6.000000,0.000000,-6.380000,0.000000,0.000000,2014-10-17 00:00:00
25%,0.000000,18.000000,2016.000000,16.000000,8.000000,0.000000,1.000000,2.000000,0.000000,0.000000,...,0.000000,0.000000,0.000000,9.000000,62.000000,0.000000,69.290000,0.000000,0.000000,2016-02-01 00:00:00
50%,0.000000,69.000000,2016.000000,28.000000,16.000000,1.000000,2.000000,2.000000,0.000000,0.000000,...,0.000000,0.000000,0.000000,14.000000,179.000000,0.000000,94.575000,0.000000,0.000000,2016-08-07 00:00:00
75%,1.000000,160.000000,2017.000000,38.000000,23.000000,2.000000,3.000000,2.000000,0.000000,0.000000,...,0.000000,0.000000,0.000000,229.000000,270.000000,0.000000,126.000000,0.000000,1.000000,2017-02-08 00:00:00
max,1.000000,737.000000,2017.000000,53.000000,31.000000,19.000000,50.000000,55.000000,10.000000,10.000000,...,26.000000,72.000000,21.000000,535.000000,543.000000,391.000000,5400.000000,8.000000,5.000000,2017-09-14 00:00:00
std,0.482918,106.863097,0.707476,13.605138,8.780829,0.998613,1.908286,0.579261,0.398561,0.097436,...,0.844336,1.497437,0.652306,110.774548,131.655015,17.594721,50.535790,0.245291,0.792798,NaN


In [15]:
# EDA - Exploratory Data Analysis
df.tail(1)

,hotel,is_canceled,lead_time,arrival_date_year,arrival_date_month,arrival_date_week_number,arrival_date_day_of_month,stays_in_weekend_nights,stays_in_week_nights,adults,...,deposit_type,agent,company,days_in_waiting_list,customer_type,adr,required_car_parking_spaces,total_of_special_requests,reservation_status,reservation_status_date
119389,City Hotel,0,205,2017,August,35,29,2,7,2,...,No Deposit,9.0,NaN,0,Transient,151.2,0,2,Check-Out,2017-09-07


In [16]:
# Data Cleaning
# Cədvəldə olan 3 sütun üzrə arrival_date_year , arrival_date_month , arrival_date_day_of_month vahid sütun yaradılır.
df["arrival_date"] = pd.to_datetime(
    df["arrival_date_year"].astype(str) + "-" +
    df["arrival_date_month"] + "-" +
    df["arrival_date_day_of_month"].astype(str),
    format="%Y-%B-%d"
)

In [17]:
# EDA - Exploratory Data Analysis
df.head(1)

,hotel,is_canceled,lead_time,arrival_date_year,arrival_date_month,arrival_date_week_number,arrival_date_day_of_month,stays_in_weekend_nights,stays_in_week_nights,adults,...,agent,company,days_in_waiting_list,customer_type,adr,required_car_parking_spaces,total_of_special_requests,reservation_status,reservation_status_date,arrival_date
0,Resort Hotel,0,342,2015,July,27,1,0,0,2,...,NaN,NaN,0,Transient,0.0,0,0,Check-Out,2015-07-01,2015-07-01


In [18]:
# EDA - Exploratory Data Analysis
# Sütundakı fərqli dəyərlərin sayı
df.nunique()

hotel                                2
is_canceled                          2
lead_time                          479
arrival_date_year                    3
arrival_date_month                  12
arrival_date_week_number            53
arrival_date_day_of_month           31
stays_in_weekend_nights             17
stays_in_week_nights                35
adults                              14
children                             5
babies                               5
meal                                 5
country                            177
market_segment                       8
distribution_channel                 5
is_repeated_guest                    2
previous_cancellations              15
previous_bookings_not_canceled      73
reserved_room_type                  10
assigned_room_type                  12
booking_changes                     21
deposit_type                         3
agent                              333
company                            352
days_in_waiting_list     

In [19]:
# Data Cleaning
# Dublikatların yoxlanması
df.duplicated().sum()

np.int64(31994)

In [20]:
# Feature Engineering
# Ümumi gecə sayı
df["total_nights"] = (
    df["stays_in_week_nights"] +
    df["stays_in_weekend_nights"]
)

# EDA - Exploratory Data Analysis
df.tail(1)

,hotel,is_canceled,lead_time,arrival_date_year,arrival_date_month,arrival_date_week_number,arrival_date_day_of_month,stays_in_weekend_nights,stays_in_week_nights,adults,...,company,days_in_waiting_list,customer_type,adr,required_car_parking_spaces,total_of_special_requests,reservation_status,reservation_status_date,arrival_date,total_nights
119389,City Hotel,0,205,2017,August,35,29,2,7,2,...,NaN,0,Transient,151.2,0,2,Check-Out,2017-09-07,2017-08-29,9


In [21]:
# Feature Engineering
# Ümumi qonaq sayı
df["total_guests"] = (
    df["adults"] +
    df["children"] +
    df["babies"]
)

# EDA - Exploratory Data Analysis
df.head(1)

,hotel,is_canceled,lead_time,arrival_date_year,arrival_date_month,arrival_date_week_number,arrival_date_day_of_month,stays_in_weekend_nights,stays_in_week_nights,adults,...,days_in_waiting_list,customer_type,adr,required_car_parking_spaces,total_of_special_requests,reservation_status,reservation_status_date,arrival_date,total_nights,total_guests
0,Resort Hotel,0,342,2015,July,27,1,0,0,2,...,0,Transient,0.0,0,0,Check-Out,2015-07-01,2015-07-01,0,2.0


In [22]:
# Feature Engineering
# Reservation ayı
df["month"] = df["arrival_date"].dt.month

# EDA - Exploratory Data Analysis
df.head(1)

,hotel,is_canceled,lead_time,arrival_date_year,arrival_date_month,arrival_date_week_number,arrival_date_day_of_month,stays_in_weekend_nights,stays_in_week_nights,adults,...,customer_type,adr,required_car_parking_spaces,total_of_special_requests,reservation_status,reservation_status_date,arrival_date,total_nights,total_guests,month
0,Resort Hotel,0,342,2015,July,27,1,0,0,2,...,Transient,0.0,0,0,Check-Out,2015-07-01,2015-07-01,0,2.0,7


In [23]:
# Feature Engineering
# Reservation həftəsi
df["weekday"] = df["arrival_date"].dt.day_name()

# EDA - Exploratory Data Analysis
df.head(1)

,hotel,is_canceled,lead_time,arrival_date_year,arrival_date_month,arrival_date_week_number,arrival_date_day_of_month,stays_in_weekend_nights,stays_in_week_nights,adults,...,adr,required_car_parking_spaces,total_of_special_requests,reservation_status,reservation_status_date,arrival_date,total_nights,total_guests,month,weekday
0,Resort Hotel,0,342,2015,July,27,1,0,0,2,...,0.0,0,0,Check-Out,2015-07-01,2015-07-01,0,2.0,7,Wednesday


In [24]:
# Insight
# Rezervasiya olunan otellərdəki miqdarlar.Ən çox rezervasiya olunan otel artıq tapılmışdır.
df["hotel"].value_counts()

hotel
City Hotel      79330
Resort Hotel    40060
Name: count, dtype: int64

In [25]:
# Insight
# Ləğv olunma və ya Cancellation faizi.
df["is_canceled"].value_counts(normalize=True)*100

is_canceled
0    62.958372
1    37.041628
Name: proportion, dtype: float64

In [26]:
# Insight
# Top 10 Ölkələr üzrə rezervasiya olunma miqdarı
df["country"].value_counts().head(10)

country
PRT    48590
GBR    12129
FRA    10415
ESP     8568
DEU     7287
ITA     3766
IRL     3375
BEL     2342
BRA     2224
NLD     2104
Name: count, dtype: int64

In [27]:
# Insight
# Ən çox rezervasiya olunan aylar
df["arrival_date_month"].value_counts()

arrival_date_month
August       13877
July         12661
May          11791
October      11160
April        11089
June         10939
September    10508
March         9794
February      8068
November      6794
December      6780
January       5929
Name: count, dtype: int64

In [28]:
# Insight
# ADR ortalaması - ADR (Average Daily Rate) — Otelin bir gecə üçün orta satış qiymətidir.
df.groupby("hotel")["adr"].mean()

hotel
City Hotel      105.304465
Resort Hotel     94.952930
Name: adr, dtype: float64

In [29]:
# EDA - Exploratory Data Analysis
df.head(1)

,hotel,is_canceled,lead_time,arrival_date_year,arrival_date_month,arrival_date_week_number,arrival_date_day_of_month,stays_in_weekend_nights,stays_in_week_nights,adults,...,adr,required_car_parking_spaces,total_of_special_requests,reservation_status,reservation_status_date,arrival_date,total_nights,total_guests,month,weekday
0,Resort Hotel,0,342,2015,July,27,1,0,0,2,...,0.0,0,0,Check-Out,2015-07-01,2015-07-01,0,2.0,7,Wednesday


In [30]:
df["total_price"] = df["total_nights"] * df["adr"]
df.head()

,hotel,is_canceled,lead_time,arrival_date_year,arrival_date_month,arrival_date_week_number,arrival_date_day_of_month,stays_in_weekend_nights,stays_in_week_nights,adults,...,required_car_parking_spaces,total_of_special_requests,reservation_status,reservation_status_date,arrival_date,total_nights,total_guests,month,weekday,total_price
0,Resort Hotel,0,342,2015,July,27,1,0,0,2,...,0,0,Check-Out,2015-07-01,2015-07-01,0,2.0,7,Wednesday,0.0
1,Resort Hotel,0,737,2015,July,27,1,0,0,2,...,0,0,Check-Out,2015-07-01,2015-07-01,0,2.0,7,Wednesday,0.0
2,Resort Hotel,0,7,2015,July,27,1,0,1,1,...,0,0,Check-Out,2015-07-02,2015-07-01,1,1.0,7,Wednesday,75.0
3,Resort Hotel,0,13,2015,July,27,1,0,1,1,...,0,0,Check-Out,2015-07-02,2015-07-01,1,1.0,7,Wednesday,75.0
4,Resort Hotel,0,14,2015,July,27,1,0,2,2,...,0,1,Check-Out,2015-07-03,2015-07-01,2,2.0,7,Wednesday,196.0


In [31]:
# profit = df[df["is_canceled"]==0].groupby("arrival_date_month")["adr"]
# Ləğv olunmayan rezervasiyalarda hər ay neçə rezervasiya olub?
df[df["is_canceled"]==0].groupby("arrival_date_month").size().sort_values()

arrival_date_month
January      4122
December     4409
November     4672
February     5372
September    6392
June         6404
April        6565
March        6645
October      6914
May          7114
July         7919
August       8638
dtype: int64

In [32]:
# Təsdiqlənmiş rezervasiyalardan ən çox gəlir gətirən ay hansıdır?
df[df["is_canceled"] == 0].groupby("arrival_date_month")["total_price"].sum().sort_values()

arrival_date_month
January       776676.84
November     1035552.96
December     1087392.29
February     1147541.48
March        1652098.59
October      1856816.72
April        2099857.48
September    2344263.36
May          2395213.64
June         2583290.92
July         4100397.07
August       4917159.06
Name: total_price, dtype: float64

In [33]:
# Ən çox təsdiq olunmuş əməliyyatı olan market seqmenti
df[df["is_canceled"] == 0].groupby("market_segment").count()

,hotel,is_canceled,lead_time,arrival_date_year,arrival_date_month,arrival_date_week_number,arrival_date_day_of_month,stays_in_weekend_nights,stays_in_week_nights,adults,...,required_car_parking_spaces,total_of_special_requests,reservation_status,reservation_status_date,arrival_date,total_nights,total_guests,month,weekday,total_price
market_segment,,,,,,,,,,,,,,,,,,,,,
Aviation,185,185,185,185,185,185,185,185,185,185,...,185,185,185,185,185,185,185,185,185,185
Complementary,646,646,646,646,646,646,646,646,646,646,...,646,646,646,646,646,646,646,646,646,646
Corporate,4303,4303,4303,4303,4303,4303,4303,4303,4303,4303,...,4303,4303,4303,4303,4303,4303,4303,4303,4303,4303
Direct,10672,10672,10672,10672,10672,10672,10672,10672,10672,10672,...,10672,10672,10672,10672,10672,10672,10672,10672,10672,10672
Groups,7714,7714,7714,7714,7714,7714,7714,7714,7714,7714,...,7714,7714,7714,7714,7714,7714,7714,7714,7714,7714
Offline TA/TO,15908,15908,15908,15908,15908,15908,15908,15908,15908,15908,...,15908,15908,15908,15908,15908,15908,15908,15908,15908,15908
Online TA,35738,35738,35738,35738,35738,35738,35738,35738,35738,35738,...,35738,35738,35738,35738,35738,35738,35738,35738,35738,35738


In [ ]:
# Müştəri tipinə görə rezervasiya sayı
df.groupby("customer_type")["total_nights"].count()

customer_type
Contract            4076
Group                577
Transient          89613
Transient-Party    25124
Name: total_nights, dtype: int64

In [ ]:
# Müştəri tipinə görə ümumi gecələmə sayı
df.groupby("customer_type")["total_nights"].sum()

customer_type
Contract            21687
Group                1663
Transient          308909
Transient-Party     76998
Name: total_nights, dtype: int64

In [ ]:
# Müştəri tipinə görə orta gecələmə müddəti
df.groupby("customer_type")["total_nights"].sum() / df.groupby("customer_type")["total_nights"].count()

customer_type
Contract           5.320658
Group              2.882149
Transient          3.447145
Transient-Party    3.064719
Name: total_nights, dtype: float64

In [ ]:
# Ləğv statusuna görə rezervasiya sayı
df.groupby("is_canceled")["lead_time"].count()

is_canceled
0    75166
1    44224
Name: lead_time, dtype: int64

In [ ]:
# Ləğv statusunun faizlə bölgüsü
df["is_canceled"].value_counts(normalize=True) * 100

is_canceled
0    62.958372
1    37.041628
Name: proportion, dtype: float64

In [ ]:
# Ləğv statusuna görə orta lead time
df.groupby("is_canceled")["lead_time"].mean()

is_canceled
0     79.984687
1    144.848815
Name: lead_time, dtype: float64

## Yekun nəticə

Bu layihədə 119,390 otel rezervasiyası üzrə məlumatları yoxladım, tarix sütunlarını emal etdim və analiz üçün yeni göstəricilər yaratdım.

* Rezervasiyaların 37.04%-i ləğv edilib.
* City Hotel 79,330 rezervasiya ilə Resort Hotel-dən daha çox rezervasiya olunub.
* Ləğv edilmiş rezervasiyalarda orta lead time 144.85 gün, ləğv edilməyən rezervasiyalarda isə 79.98 gün olub. Bu nəticə daha erkən edilən rezervasiyalarda ləğv səviyyəsinin daha yüksək ola biləcəyini göstərir, lakin birbaşa səbəb əlaqəsini sübut etmir.
* City Hotel üzrə orta ADR 105.30, Resort Hotel üzrə isə 94.95 olub.
* Təsdiqlənmiş rezervasiyaların sayı və hesablanmış rezervasiya dəyəri ən yüksək avqust ayında qeydə alınıb.
* Contract müştəri tipində orta gecələmə müddəti 5.32 gecə ilə digər müştəri tiplərindən daha yüksək olub.

## Məlumatların emalı ilə bağlı qeyd

Yalnız bütün sütunları boş olan sətirləri silmək üçün `dropna(how="all")` yanaşmasından istifadə etdim. `agent` və `company` sütunlarındakı boş dəyərləri silmədim, çünki həmin sətirlərdə rezervasiyaya aid digər məlumatlar mənalıdır. Bu boşluqlar rezervasiyanın agent və ya şirkət vasitəsilə edilmədiyini göstərə bilər; buna görə sətirlərin tam silinməsi analizin nəticələrini təhrif edə bilərdi.

`total_price` göstəricisini ADR ilə ümumi gecələmə sayının hasili kimi hesabladım. Buna görə bu göstərici rəsmi gəlir deyil, dataset əsasında hesablanmış təxmini rezervasiya dəyəridir.

Dataset təlim və analiz məqsədi daşıyır. Nəticələr real otel şirkətinin cari fəaliyyət göstəriciləri kimi qəbul edilməməlidir.
